agent-common/src/com/thoughtworks/go/agent/launcher/ServerBinaryDownloader.java

```diff
@@ -17,161 +17,134 @@
+            this.sslPort = response.getFirstHeader(SSL_PORT_HEADER).getValue();
         }
-        return headers;
     }
 
-    private String checkUrl() {
+    protected boolean download(final DownloadableFile downloadableFile) throws Exception {
+        File toDownload = downloadableFile.getLocalFile();
+        LOG.info("download of " + toDownload + " started at " + new Date());
         String url = downloadableFile.url(urlGenerator);
-        try {
-            new URL(url);
-        } catch (MalformedURLException mue) {
-            throw new RuntimeException(
-                    "URL you provided to access Go Server: " + downloadableFile.url(urlGenerator) + " is not valid");
-        }
-        return url;
-    }
+        final HttpRequestBase request = new HttpGet(url);
+        request.setConfig(RequestConfig.custom().setConnectTimeout(HTTP_TIMEOUT_IN_MILLISECONDS).build());
 
-    private static boolean checksOut(File file, String expectedSignature) {
-        try (FileInputStream input = new FileInputStream(file)) {
-            FileDigester fileDigester = new FileDigester(input, new NullOutputStream());
-            fileDigester.copy();
-            return expectedSignature.equals(fileDigester.md5());
-        } catch (Exception e) {
-            throw new RuntimeException(e);
+        try (CloseableHttpClient httpClient = httpClientBuilder.build();
+             CloseableHttpResponse response = httpClient.execute(request)) {
+            LOG.info("got server response at " + new Date());
+            handleInvalidResponse(response, url);
+            response.getEntity().writeTo(new FileOutputStream(downloadableFile.getLocalFile()));
+            LOG.info("piped the stream to " + downloadableFile + " at " + new Date());
         }
+        return true;
     }
 
-    private boolean download() throws Exception {
-        InputStream body = null;
-        File toDownload = new File(downloadableFile.getLocalFileName());
-        try (OutputStream outputFile = new BufferedOutputStream(new FileOutputStream(toDownload))) {
-            LOG.info("download of " + toDownload + " started at " + new Date());
-            ServerCall.ServerResponseWrapper invoke = serverCall.invoke(new HttpGet(checkUrl()), rootCertFile, sslVerificationMode);
-            body = invoke.body;
-            LOG.info("got server response at " + new Date());
-            IOUtils.copy(body, outputFile);
-            LOG.info("pipe the stream to " + downloadableFile + " at " + new Date());
-            return true;
-        } catch (Exception e) {
-            String message = "Couldn't access Go Server with base url: " + downloadableFile.url(urlGenerator) + ": " + e.toString();
-            LOG.error(message);
-            throw new Exception(message, e);
-        } finally {
-            IOUtils.closeQuietly(body);
+    private void handleInvalidResponse(HttpResponse response, String url) throws IOException {
+        try (StringWriter sw = new StringWriter();
```

Comment:
The ```StringWriter``` is in there because ```PrintWriter``` has a close implementation. So it must in scope of the ```PrintWriter```. I guess that's the reason why it's marked with the ```Closeable``` interface